Render a FailureComponent when unauthorized

mehiel opened the pull request:
With this change we introduce a denyStrategy property with enumerated
values REDIRECT and NULLIFY to enable the option of not rendering a
component instead of redirecting.

USE CASE:
This is helpful for example if the user wants to use mostly the same
set of rules for route authorization and rendering the correct menu
items for a role. We can wrap MenuItems as we wrap route components
but in MenuItems we want to render only those accessible instead of
redirecting. Another case could be dashboard widgets or actions per
role.

mehiel commented:
@mjrussell I'm not sure about the API and the name of the param (`denyStrategy`) but that's something we can discuss on if you're interested on this functionality. Keep it up!

mjrussell commented:
@mehiel This is great! 

Thank you for contributing, I can definitely see a benefit of using the auth wrapper in this way. I'm planning on taking a little bit more time to look at this later today & tomorrow.

mehiel commented:
@mjrussell that would be great. Some insights:

1. I wrote this patch on v0.5.0 so its been used internally for about 4 months now and it served its purpose really well. Of course I merged the latest master and edited it before submitting the PR
2. The really fun part is that I started working to your code to add this feature and after half an hour I discovered that it was already there, we just had to enable it (by making redirection optional)
3. When I wrote that patch I though that a boolean may not be enough in case we'd like to add more "strategies" in the feature to handle non-authorization to the child component but now I'm not sure about it. The truth is that I added no more strategies and these two are more than enough even for a project with complex permissions, roles inheritance etc.. I'd like your thoughts on that. I don't want that be overengineered
4. Apologies for not providing the tests already, if we decide on the API I'll update this PR accordingly
5. The feature of having failure path optionally a function, introduced in 0.6.0, also helped me implement some complex rules easily in combination with this one

mjrussell commented:
@mehiel 

> When I wrote that patch I though that a boolean may not be enough in case we'd like to add more "strategies" in the feature to handle non-authorization to the child component but now I'm not sure about it

Yeah I was thinking the same thing while reading this over. I also am wondering about if we should make it more general than "nullify" essentially a boolean logic of rendering a component based on auth. We could make it render in the following cases:

1. If authN and authZ - render wrapped Component
2. If authenticating/authorizing - render LoadingComponent
3. If not authN/authZ and redirect (default) - render nothing/empty
4. If not authN/authZ and no redirect - render FailureComponent (defaults to null)

In this way maybe you'd have a menu item that you wanted instead of removing it, you instead render a span that says you aren't allowed to use this feature.

I do like the name of strategy, but I do wonder if a simpler API with a boolean and flexible failure case would be easier to understand and a bit more powerful.

We could probably also slightly tweak the component if they aren't redirecting (like disallow the router proptypes and the onEnter) instead of removing the `isRequired` completely for all cases.

mjrussell commented:
And don't worry about the tests! Thats easy once we get the rest of it settled

mehiel commented:
@mjrussell I like that. The simplest approach seems to be:

If user sets failureRedirectPath prop then redirection is going to happen and it takes precedence. Otherwise it's rendering that takes control with null as the default FailureComponent and you can specify a custom one if you want to.

> We could probably also slightly tweak the component if they aren't redirecting (like disallow the router proptypes and the onEnter) instead of removing the isRequired completely for all cases.

Sure, that's better. If we agree on the above API I could update this PR and we can review it again.

mehiel commented:
BUT the simplest (TM) scenario above breaks compatibility :-( We'd need proper versioning and communication for such a change..

mjrussell commented:
Was just going to comment on breaking backwards compatibility haha. I'd prefer to try to come up with something like what you described, but backwards compatible. I want to do a few more versions before 1.0.

Would setting the failureRedirectPath to `null` be too hacky? Could also export some sort of symbol so like:

```js
import UserAuthWrapper, { NO_REDIR } from 'redux-auth-wrapper'

UserAuthWrapper({
  failureRedirectPath: NO_REDIR,
})
```

mehiel commented:
It's not that intuitive for sure but not hacky. The other way to do it is if you define a FailureComponent with no failureRedirectPath then you have it.

```
UserAuthWrapper({
   failureComponent: null
})
```

mjrussell commented:
> The other way to do it is if you define a FailureComponent with no failureRedirectPath then you have it.

I think we want to allow FailureComponent to just be set to null by default (meaning don't render anything). 

Im leaning towards the setting the failurePath as null or an exported constant. 

Or even an alternative Wrapper which just sets some params for the the other one. Would be nice if the existing one was like `UserAuthRedirect` whereas a new one could be like `UserAuthRender` or something like that. Then we could clean up the API to not include any routing stuff.

mehiel commented:
Ok. Decision Time :8ball:. 

From what you proposed I'd prefer having 2 components more than setting failurePath to null. But I'm fine with both so it's your call.

mjrussell commented:
@mehiel yeah agree lets get something done and can always tweak it as we use it more.

So you are suggesting we go with `FailureComponent` and `redirect` properties? Im happy with that

mjrussell commented:
@mehiel any update on this? Are you waiting on me for anything?

rewop commented:
I like the approach of `FailureComponent` as it is in line with the `LoadingComponent`. 

To keep this simple, I believe a good way is to give more precedence is to add a `FailureComponent` property that has high precedence than `failureRedirectPath`. 

We can use the following strategy in case of failure to preserve backwards compatibility and allow nullify.

```javascript
if (typeof FailureComponent === "undefined") // do redirect
else if (FailureComponent === null) // nullify
else // render FailureComponent
```

Just few notes toward version 1.0.

- I believe the api could be improved a lot, to remove some hard dependencies, such has `react-redux` (I am using [apollo-react](https://github.com/apollostack/react-apollo) that wraps the `AuthWrapper`)

- I would extract general parts and have a more functional approach. This would allow composition, and to use the api also outside redux scope.

mjrussell commented:
> To keep this simple, I believe a good way is to give more precedence is to add a FailureComponent property that has high precedence than failureRedirectPath.

Yeah I like this too, simple and intuitive.

> I believe the api could be improved a lot, to remove some hard dependencies, such has react-redux

Interesting...I haven't thought much about this. I think my primary goal is to have it be easy for any redux user to use, but I could see being able to override the "provider" system for more advanced users.

> I would extract general parts and have a more functional approach. This would allow composition, and to use the api also outside redux scope

👍 Its grown very organically so far so it could definitely use some refactoring towards composibility

mehiel commented:
Hi all. I'm out of office till next week, thus the delay. 

@mjrussell no it's pretty clear how we'll do this now, I update the PR in the next few days.

> >To keep this simple, I believe a good way is to give more precedence is to add a FailureComponent property that has high precedence than failureRedirectPath.

> Yeah I like this too, simple and intuitive.

Great. I like that too. And it's totally weird that we couldn't see that :( Thanks @rewop.

About removing some hard dependencies, I see that coming as well, as the project evolves. I find myself looking for redux-less architectures already and at the same time it should be relatively easy for this project to abstract it away from redux and leave it there as an option.

mjrussell commented:
> Hi all. I'm out of office till next week, thus the delay.

No problem! Just wanted to double check. Thanks for working this. Im going to release a new version today and we will get this in the next one

mehiel commented:
Hi @mjrussell,
 back from vacations. I'll update this PR later today and we can further review its status.

mehiel commented:
@mjrussell, @rewop 
PR updated. Please review and if everything looks good I'll update the tests.

FYI I rewrote the PR commit to keep git history clean, hope you don't mind.

mehiel commented:
Do you thing we should set `onEnter` only when `isRedirect`?

mjrussell commented on src/factory.js:
```diff
@@ -92,11 +96,16 @@ export default function factory(React, empty) {
+              (wasAuthorized && !willBeAuthorized) ||
+              // 2. Was not authorized and authenticating
+              (wasAuthenticating && !willBeAuthorized)
             ) {
```
Seems like would be cleaner if we did 
```js
          // Redirect if:
          if ( isRedirect &&
               // 1. Was authorized, but no longer
              (wasAuthorized && !willBeAuthorized) ||
              // 2. Was not authorized and authenticating
              (wasAuthenticating && !willBeAuthorized)
            ) {
            createRedirect(nextProps.location, this.getRedirectFunc(nextProps), nextProps.failureRedirectPath)
```

mehiel replied:
Is this correct? Seems like we need an extra parenthesis wrapping the `||` expression.

mjrussell replied:
Oops good catch

mehiel replied:
It's getting a bit complex like that. Am I missing something?

```
if ( 
  isRedirect &&
  // 1. Not currently authenticating AND
  !willbeAuthenticating &&
  (
    // 2a. Was authorized, but no longer OR
    (wasAuthorized && !willBeAuthorized) ||
    // 2b. Was not authorized and authenticating
    (wasAuthenticating && !willBeAuthorized)
  )
) {
```

mjrussell replied:
Oh I see what you are doing now, sorry. I missed that you pulled the `!willBeAuthenticating` out in the original PR. We can leave it as you had it then. I don't love the return nothing from an if but it makes the second part cleaner

mehiel replied:
I don't like early `return`ing either but it looked a bit cleaner here.

mjrussell commented on src/factory.js:
```diff
@@ -124,8 +133,10 @@ export default function factory(React, empty) {
-            return React.createElement(empty)
+            // Unauthorized: Display FailureComponent or nothing
+            // if FailureComponent is undefined the user will never see
+            // this because she'll be redirected to failureRedirectPath
```
replace `she'll` with `they will`

mjrussell commented:
Looks good, some minor feedback.

> Do you thing we should set onEnter only when isRedirect?

Yeah definitely if there is a clean way to do it.

mjrussell commented:
I just pushed to master a commit that upgrades and pins dependencies. The build was failing with some annoying react and react native version mismatches. Might want to rebase off it again

mehiel commented:
> Might want to rebase off it again

done.

mjrussell commented on src/factory.js:
```diff
@@ -39,6 +40,12 @@ export default function factory(React, empty) {
       })
     }
 
+    const isRedirect = FailureComponent !== undefined
```
I think this should actually be `FailureComponent === undefined` right?

mehiel replied:
OMG. This is at least embarrassing :-) Fixing it asap. 

It started as `isNotRedirect` but after that I found out that the complement is easier to read. Well..

mjrussell replied:
Haha I was trying to understand why after created a backwards compatible API all the tests were failing. Refactoring is the source of so many bugs 😄

mehiel replied:
+1 on github for hiding comments on outdated diffs :dancer: 
Fix pushed.

mjrussell commented:
Awesome, now just for some test coverage.

Probably test for:
- No redirect when setting FailureComponent to an actual component (or set to null)
  - Both on mount and willReceiveProps shouldnt trigger a redirect
- No onEnter when FailureComponent set

mehiel commented:
kk. I'll come back to that before the weekend.

mehiel commented:
@mjrussell tests added. I have to somehow increase the coverage though.

mjrussell commented:
Yeah there is an extra case where we should check when the props change (triggering componentWillReceiveProps) that it wont trigger a redirect if there is a FailureComponent. So start with logged in, and dispatch the logout with a failure.

mehiel commented:
Cool, thanks @mjrussell .. Updates pushed. I just run the `test:cov` myself and saw the path that wasn't covered.

FYI I also renamed that `isRedirect` to `shouldRedirect` which reads better.

mjrussell commented:
Sweet, ok one last thing!

It seems like FailureComponent is a node (something to be rendered) instead of actually a component like LoadingComponent. What do you think about changing it to a React Component? That was if we need to pass down the auth data (for instance maybe to show a reason of why it failed) or other props we could do that.

mjrussell commented:
Sorry for all the back and forth on this, just noticed that when looking at the tests

mehiel commented:
No prob at all.

How are we supposed to change it to a React Component? Passing `authData` seems legit.

mehiel commented:
You probably mean to `<FailureComponent {...whateverProps} />`, right?

mjrussell commented:
> You probably mean to `<FailureComponent {...whateverProps} />`, right?

Yup exactly, only catch is to return null if its set to null.

Probably can pass `...otherProps` and `authData` to it.

mehiel commented:
lol was about to comment:

`return <FailureComponent authData={authData} {...otherProps} /> || React.createElement(empty)`

Yeah I'll check `null` specifically.

mehiel commented:
After merging this one I could probably easily send a PR for #75 as well. But I think it shouldn't be part of this one for clarity, even though it's kind of trivial.

mjrussell commented on src/factory.js:
```diff
@@ -124,21 +133,25 @@ export default function factory(React, empty) {
+            // Unauthorized: Display FailureComponent or nothing
+            // if FailureComponent is undefined the user will never see
+            // this because they will be redirected to failureRedirectPath
+            return FailureComponent ? <FailureComponent authData={authData} {...otherProps} /> : React.createElement(empty)
```
one catch here, I think empty isn't actually set to be null. Its a [div](https://github.com/mjrussell/redux-auth-wrapper/blob/master/src/index.js#L4) 😕  Seems like a `<div />` might trip people up if they expect it to be hidden. Should we change the default empty?

I believe I used a div initially because React Router can get angry if you use null. I can verify quickly though

mehiel replied:
Hmmm.. I tried to avoid extra conditions in there but in that case I should add null explicitly. You're probably right, having an empty `<div>` when we should have nothing is probably wrong. It can break the layout in some cased and leak information in others.

Not sure about React Router's behavior in that case. Never being there.

mjrussell replied:
Agree, the block layout can definitely cause issues. 

Can you add an extra case for now, and I'll follow up on seeing if I can replace the `<div/>` with nulls after

mehiel replied:
@mjrussell check the update. I added one more `else if` to avoid nesting conditions.

mjrussell replied:
Am I missing something? I dont see a change in your last push unless github diff is being slow

mehiel replied:
pfff.. its getting late at GMT+3 :)

mjrussell commented:
All set! Thanks for all your work on this, I think this is going to be a really appreciated feature. I'll probably try to get #76 and #75 done and add a little documentation and then cut a new release.

mehiel commented:
It's been a pleasure. Keep it up!

mjrussell commented:
Just released version 0.8! It includes the FailureComponent
